gh-badges/lib/badge-renderers.js

```diff
@@ -0,0 +1,596 @@
+'use strict'
+
+const anafanafo = require('anafanafo')
+
+const fontFamily = 'font-family="DejaVu Sans,Verdana,Geneva,sans-serif"'
+const socialFontFamily =
+  'font-family="Helvetica Neue,Helvetica,Arial,sans-serif"'
+
+function capitalize(s) {
+  return `${s.charAt(0).toUpperCase()}${s.slice(1)}`
+}
+
+function escapeXml(s) {
+  if (s === undefined || typeof s !== 'string') {
+    return undefined
+  } else {
+    return s
+      .replace(/&/g, '&amp;')
+      .replace(/</g, '&lt;')
+      .replace(/>/g, '&gt;')
+      .replace(/"/g, '&quot;')
+      .replace(/'/g, '&apos;')
+  }
+}
+
+function roundUpToOdd(val) {
+  // Increase chances of pixel grid alignment.
+  return val % 2 === 0 ? val + 1 : val
+}
+
+function preferredWidthOf(str) {
+  return roundUpToOdd((anafanafo(str) / 10) | 0)
+}
+
+function computeWidths({ label, message }) {
+  return {
+    labelWidth: preferredWidthOf(label),
+    messageWidth: preferredWidthOf(message),
+  }
+}
+
+function renderLogo({
+  logo,
+  logoWidth = 14,
+  logoPadding = 0,
+  extraPadding = 0,
+}) {
+  if (logo) {
+    const x = 5 + extraPadding
```

Comment:
For my own edificiation, where does the extra +5 and +3 come from/why those specific values?

Reply:
Broadly because they were hard-coded values in the original templates we're trying to match.

I can't link to a line here because GitHub won't render the files, but if you look at say line 8 of
https://github.com/badges/shields/blob/v3-dev/gh-badges/templates/flat-square-template.svg
the X and Y values for the image are just hard-coded to 5 and 3. This function generalises so that we can pass additional offsets for `for-the-badge` style.

Any thoughts on whether there is a more intuitive way to express it?

Reply:
> Any thoughts on whether there is a more intuitive way to express it?

Not really, other than perhaps an inline comment or something like that. I hadn't really looked at the svg template files before so when perusing the js here they seemed like semi-magical numbers

Reply:
Maybe instead of assuming a base position and allowing it to be modified with an offset, we just explicitly pass in X and Y for each badge size, then we can make it more obvious how the positions relate to each badge format. That might be clearer..

Reply:
That seems like a nice idea. Also, we could call it `horizPadding` even if it’s an `x` value.

Reply:
cool. I'll aim to push a change for this tomorrow. Having chewed it over a bit more, I've actually realised there is an error I never noticed before.

We hard-code y=3 for `flat` and `plastic`, but y should really be derived from the badge height, so it should be ([badge height] - [logo height])/2 so for the flat badge that is (20 - 14)/2 = 3, but then the plastic badge is slightly shorter which means when we hard-code the y pos to 3, the logo actually renders a bit too low vertically i.e:

https://img.shields.io/badge/plastic-style-blue?style=plastic&logo=javascript
<img src="https://img.shields.io/badge/plastic-style-blue?style=plastic&logo=javascript" width="400%" height="400%" />

Reply:
done in c5ef23a